qubes/ext/utils.py

```diff
@@ -100,3 +159,56 @@ def compare_device_cache(vm, devices_cache, current_devices):
             if cached_front is not None:
                 detached[dev_id] = cached_front
     return added, attached, detached, removed
+
+
+async def confirm_device_attachment(device, frontends) -> str:
+    try:
+        return await _do_confirm_device_attachment(device, frontends)
+    except Exception as exc:
+        print(str(exc.__class__.__name__) + ":", str(exc), file=sys.stderr)
+        return ""
+
+
+async def _do_confirm_device_attachment(device, frontends):
+    socket = "device-agent.GUI"
+
+    app = tuple(frontends.keys())[0].app
+    doms = app.domains
+
+    front_names = [f.name for f in frontends.keys()]
+
+    try:
+        guivm = doms["dom0"].guivm.name
+    except AttributeError:
+        guivm = "dom0"
+
+    number_of_targets = len(front_names)
+
+    params = {
+        "source": device.backend_domain.name,
+        "device_name": device.description,
+        "argument": device.port_id,
+        "targets": front_names,
+        "default_target": front_names[0] if number_of_targets == 1 else "",
+        "icons": {
+            (
+                dom.name if dom.klass != "DispVM" else f"@dispvm:{dom.name}"
+            ): dom.icon
+            for dom in doms.values()
+        },
+    }
+
+    socked_call = asyncio.create_task(
+        call_socket_service(guivm, socket, "dom0", params, SOCKET_PATH)
+    )
+
+    while not socked_call.done():
+        await asyncio.sleep(0.1)
```

Comment:
Why this loop? the `await socked_call` below should be enough, no?

Reply:
If we don't wait for the task to complete, using await in this case results in tasks being destroyed (including resolve_conflicts_and_attach) so attachment is skipped. I'm not entirely sure what causes this.

Reply:
Weird, I'll try to find out what is going on.

Reply:
I've tried and it works for me just fine without this part. I've tried by assigning block device with ask, and then connecting and disconnecting it - I got the prompt and when accepted the device got attached.

BTW, attaching device could use some log entry, especially when automatic or with ask (so you have some trail to which qube it got attached).

Reply:
Ok, I got once "Task was destroyed but it is pending". Skipping create task (`ask_response = await call_socket_service(...)`)helped.

Reply:
in my case skipping creating task always ended up with `Task was destroyed but it is pending`